### PR TITLE
Rush utility testing and error handling

### PR DESCRIPTION
**Pull request by ecraig12345**
Add tests for the rush-related utilities, and improve their error handling. Some error handling improvements for the upgrade repo task too, but it still needs more work.

### CODE REVIEW

File: packages/just-scripts/src/tasks/upgradeRepoTask.ts [hunk 1]
```diff
@@ -5,52 +5,55 @@ import {
...
+  outdatedStacks.forEach(stack => {
+    const { dependencies = {}, devDependencies = {} } = scriptsPackageJson;
+    if (devDependencies[stack.name]) {
+      devDependencies[stack.name] = options.latest ? `^${stack.latest}` : `^${stack.wanted}`;
+    } else if (dependencies[stack.name]) {
+      dependencies[stack.name] = options.latest ? `^${stack.latest}` : `^${stack.wanted}`;
+    }
+  });
+
+  fse.writeJsonSync(scriptsPackageJsonPath, scriptsPackageJson, { spaces: 2 });
```
Comment (kenotron) [inline packages/just-scripts/src/tasks/upgradeRepoTask.ts]: I think this way we would potentially write blank objects out for deps and devDeps, right?

File: packages/just-scripts/src/tasks/upgradeRepoTask.ts [hunk 2]
```diff
@@ -5,52 +5,55 @@ import {
...
-
-    const rushConfig = readRushJson(rootPath);
-
-    // uses Array.reduce to sequentially loop through promise
-    rushConfig.projects.reduce(async (previousPromise: Promise<void>, project: any) => {
-      await previousPromise;
-      const projectPath = path.join(rootPath, project.projectFolder);
-      return upgradeStackPackageJsonFile(projectPath, path.join(scriptsPath, 'node_modules'));
-    }, Promise.resolve());
+  for (const project of rushConfig.projects) {
```
Comment (kenotron) [inline packages/just-scripts/src/tasks/upgradeRepoTask.ts]: I don't think this works to serially do the work for you. The reduce syntax is probably your best bet to get this done serially. The async pattern you are looking for is called "async iterator":

https://nodejs.org/download/nightly/v10.0.0-nightly2018033083d44bee01/docs/api/stream.html#stream_readable_asynciterator

It's coming but it's a little strange:

```
for await (let project of rushConfig.projects) { ... }
```

File: packages/just-scripts/src/tasks/upgradeRepoTask.ts (hunk 1 above)
Comment (ecraig12345) [reply packages/just-scripts/src/tasks/upgradeRepoTask.ts]: Yeah, but I don't think that's a problem?

File: packages/just-scripts/src/tasks/upgradeRepoTask.ts (hunk 2 above)
Comment (ecraig12345) [reply packages/just-scripts/src/tasks/upgradeRepoTask.ts]: The reduce syntax is super unreadable... If what I did doesn't work properly, I think this is better (and does execute serially, unlike the incorrect version I wrote in another PR at some point).
```
let promise = Promise.resolve();
for (const thing of things) {
  promise = promise.then(() => doStuff(thing));
}
return promise;
```

Comment (ecraig12345) [reply packages/just-scripts/src/tasks/upgradeRepoTask.ts]: Fixed to use for loop without await

**Pull request merged**